### PR TITLE
feat(pubsublite): Pub/Sub Lite admin client

### PR DESCRIPTION
**Pull request by tmdiep**
Implements pubsublite.Client, which wraps the Pub/Sub Lite Admin Service.
Includes integration tests for admin operations.

### CODE REVIEW

**Comment by tmdiep** [comment]
Unit tests will be in a future PR, as they would make this one too large and I figured integration tests are more worthwhile.

File: pubsublite/pubsublite.go
Comment (codyoss) [inline pubsublite/pubsublite.go:50]: Are nil configs valid for these admin operations? Should this be passed by value?

File: pubsublite/pubsublite.go
Comment (codyoss) [inline pubsublite/pubsublite.go:65]: Is there a way to use the same struct as the one used for creation?

File: pubsublite/integration_test.go
Comment (tmdiep) [reply pubsublite/integration_test.go]: Done, this looks much nicer.

File: pubsublite/integration_test.go
Comment (tmdiep) [reply pubsublite/integration_test.go]: Done

File: pubsublite/pubsublite.go
Comment (tmdiep) [reply pubsublite/pubsublite.go:42]: Ah, thanks for catching. I completely forgot!

File: pubsublite/integration_test.go
Comment (tmdiep) [reply pubsublite/integration_test.go:183]: Ack, we'll definitely need to expand on doc.go eventually.

File: pubsublite/integration_test.go
Comment (tmdiep) [reply pubsublite/integration_test.go]: Done, good point

File: pubsublite/pubsublite.go
Comment (tmdiep) [reply pubsublite/pubsublite.go:28]: Hmm, I might leave this as Client (and changed NewAdminClient to NewClient) to match pubsub.Client. One of our goals was to make the pubsublite interface as consistent as possible with pubsub. Given that we've had to make quite a few deviations already, we could live with this one. (If starting from scratch, I would actually call this AdminClient too.)

File: pubsublite/pubsublite.go
Comment (tmdiep) [reply pubsublite/pubsublite.go]: Removed. True, it's not needed internally and only added for convenience for users.

File: pubsublite/pubsublite.go
Comment (tmdiep) [reply pubsublite/pubsublite.go]: Done

File: pubsublite/pubsublite.go
Comment (tmdiep) [reply pubsublite/pubsublite.go]: It was exposed in case users wanted to check their inputs, but I changed this to internal for now.

File: pubsublite/pubsublite.go
Comment (tmdiep) [reply pubsublite/pubsublite.go]: Yes, they are required. Changed to value.

File: pubsublite/pubsublite.go
Comment (tmdiep) [reply pubsublite/pubsublite.go]: I guess we could, and add comments to the fields that cannot be updated after creation. I added these structs for a couple of reasons:

- To follow the *ToUpdate convention seen in many of the other modules. 
- For future flexibility, as some fields may be tri-state (e.g. for string: don't update/set to empty string/set to non-empty string). We actually almost had this issue with RetentionDuration until we decided to differentiate between unset/zero/non-zero duration.

File: pubsublite/pubsublite.go
Comment (tmdiep) [reply pubsublite/pubsublite.go:28]: @hongalex, do you have a strong preference?

Comment (hongalex) [reply pubsublite/pubsublite.go:28]: Considering we'll using separate publisher/subscriber clients for Lite, I think the deviation makes sense. Keeping it the same might confuse new users anyway since Pub/Sub uses the generic `Client` struct for everything, something that Lite won't do.

File: pubsublite/pubsublite.go
Comment (hongalex) [reply pubsublite/pubsublite.go]: Agreed, the `toUpdate` pattern is used to facilitate field masking for updates for pub/sub and should be kept for Lite as well.

File: pubsublite/integration_test.go
Comment (tmdiep) [reply pubsublite/integration_test.go:183]: On the subject of infinite retention, I would appreciate a second opinion on something. More context in b/170457893#comment17.

So for topic creation, I treated TopicConfig.RetentionDuration = {InfiniteRetention|nil} to behave exactly the same, i.e. infinite retention. For topic update, TopicConfigToUpdate.RetentionDuration = nil means leave the retention as-is; = InfiniteRetention means to clear the current retention value and make it infinite.

WDYT? Is this confusing?

Alternatively, TopicConfig.RetentionDuration could be changed to a time.Duration, where an unset duration proto is represented as InfiniteRetention, rather than nil. This makes the semantics identical in TopicConfig and TopicConfigToUpdate.

File: pubsublite/pubsublite.go
Comment (tmdiep) [reply pubsublite/pubsublite.go]: Seems no one is particularly attached to `Client` :)
I've renamed to AdminClient and the file to admin.go.

File: pubsublite/integration_test.go
Comment (tmdiep) [reply pubsublite/integration_test.go:183]: > Alternatively, TopicConfig.RetentionDuration could be changed to a time.Duration, where an unset duration proto is represented as InfiniteRetention, rather than nil. This makes the semantics identical in TopicConfig and TopicConfigToUpdate.

I just went ahead and implemented this, as I think having the same semantics may be less confusing. Can change it back if anyone likes the original implementation.

**Comment by tmdiep** [comment]
Thanks both, all comments addressed.
But I'd like to get Manu to review as well, once it's possible!

**Comment by tmdiep** [comment]
Manu will be OOO the next couple of days, so I'll get this committed to unblock dependent PRs. This is ready for approval and merge now.

**Review by hongalex — APPROVED** [review]

File: pubsublite/config.go
Comment (codyoss) [inline pubsublite/config.go:60]: Is it valid to have a PerPartitionBytes of zero and RetentionDuration of zero?

File: pubsublite/admin.go
Comment (codyoss) [inline pubsublite/admin.go:75]: Here, and with other methods, XXXPath does not seems like a meaningful name. It makes sense if you know that it is used to build a resource name, but if you don't know that the name is confusing in my opinion. I wonder if there is a way to refactor this so that all you need to pass to the method is a  context and `TopicID`.

**Review by codyoss — COMMENTED** [review]
No need to block on these as more iteration can happen later. Just some thoughts.

File: pubsublite/config.go
Comment (tmdiep) [reply pubsublite/config.go:60]: Nope, good catch. Updated the comments.

File: pubsublite/admin.go
Comment (tmdiep) [reply pubsublite/admin.go:75]: The topic is uniquely identified by its {project, zone, topic id}. I didn't store the project and zone in the client because we can have cross-project subscriptions, where the topic and subscription resources are not necessarily owned by the same project. So having a default project set for the client could cause some confusion for, e.g. Create/Update Subscription (i.e. is the default project used for the topic or the subscription?).

I also saw that [pubsub.Client](https://godoc.org/cloud.google.com/go/pubsub#pkg-index) had evolved to support multiple variants Topic/TopicInProject, Subscription/SubscriptionInProject, so wanted to avoid this, as it seems multi-project admin does occur.

Similar remarks for zone, as the admin server is regional. Users can use the same client to manage all resources within the region.

For these reasons, I thought it would be best to give the client more flexibility with projects and zones; avoiding the multiple variants that pubsub.Client has.

Comment (tmdiep) [reply pubsublite/admin.go:75]: I forgot to mention: we have planned regional replication for topics (b/171412217), so the admin client could end up dealing with multiple zones within the same region, so storing a default zone could cause confusion and/or proliferation of method variants.

**Pull request merged**
